apps/emqx_gateway_mqttsn/src/emqx_mqttsn_channel.erl

```diff
@@ -461,48 +460,61 @@ handle_in(?SN_ADVERTISE_MSG(_GwId, _Radius), Channel) ->
     % ignore
     shutdown(normal, Channel);
 handle_in(
-    ?SN_PUBLISH_MSG(
-        #mqtt_sn_flags{
-            qos = ?QOS_NEG1,
-            topic_id_type = TopicIdType
-        },
-        TopicId,
-        _MsgId,
-        Data
-    ),
+    Publish =
+        ?SN_PUBLISH_MSG(
+            #mqtt_sn_flags{
+                qos = ?QOS_NEG1,
+                topic_id_type = TopicIdType
+            },
+            TopicId,
+            MsgId,
+            Data
+        ),
     Channel = #channel{conn_state = idle, registry = Registry}
 ) ->
-    %% FIXME: check enable_qos3 ??
-    TopicName =
-        case (TopicIdType =:= ?SN_SHORT_TOPIC) of
-            true ->
-                <<TopicId:16>>;
-            false ->
-                emqx_mqttsn_registry:lookup_topic(
-                    Registry,
-                    ?NEG_QOS_CLIENT_ID,
-                    TopicId
-                )
-        end,
-    _ =
-        case TopicName =/= undefined of
-            true ->
-                Msg = emqx_message:make(
-                    ?NEG_QOS_CLIENT_ID,
-                    ?QOS_0,
-                    TopicName,
-                    Data
-                ),
-                emqx_broker:publish(Msg);
-            false ->
-                ok
-        end,
-    ?SLOG(debug, #{
-        msg => "receive_qo3_message_in_idle_mode",
-        topic => TopicName,
-        data => Data
-    }),
-    {ok, Channel};
+    case check_negative_qos_enable(Publish, Channel) of
+        ok ->
+            TopicName =
+                case TopicIdType of
+                    ?SN_SHORT_TOPIC ->
+                        TopicId;
```

Comment:
Not sure if this is intended, is `TopicId` an integer or not?

Reply:
The `TopicId` field can be a short topic name or a pre-defined ID according to the specification.
<img width="879" alt="image" src="https://github.com/emqx/emqx/assets/13825269/7954abf6-b35a-499d-96d3-99b7b380355f">

It should be a 2-byte string; it is actually parsed out here:
https://github.com/emqx/emqx/blob/273dedf7a6bc6c6913be6365d24a602b42f71f77/apps/emqx_gateway_mqttsn/src/emqx_mqttsn_frame.erl#LL153C4-L153C4